File: basic.learning.js
// frappe.ui.form.on('Client Side Scripting', {
    
//   refresh:function(frm){
//       frm.add_custom_button('Click me button',()=>{
//           frappe.msgprint("You clicked me");
//       })
      
//       frm.add_custom_button('Click button 1',()=>{
//           frappe.msgprint("You clicked me");
//       },"click me") // click me is button name
      
//       frm.add_custom_button('Click button 2',()=>{
//           frappe.msgprint("You clicked me");
//       },"click me") // click me is button name
      
//   },
    
//         validate:function(frm) {
//         frm.set_value('full_name', `${cur_frm.doc.first_name} ${cur_frm.doc.middle_name} ${cur_frm.doc.last_name}` );
// },
    
//     validate:function(frm) {
//       const arrayOfMembers=cur_frm.doc.family_members   
//       console.log(arrayOfMembers);
      
//     let totalSalary =0;
//     arrayOfMembers.forEach(member =>{
//         totalSalary += member.salary
//     })
//      console.log(totalSalary);
//     frm.set_value('total_salary', totalSalary);
//     // frappe.msgprint("kindly check the total salary of the family before submiting")
//     }
// });



    
    
//     validate:function(frm) {
//         console.log(cur_frm.doc.age);
//         frm.set_value('email', 'samuvel7860@gmail.com')
//         for (var i = 0; i < cur_frm.doc.family_members.length; i++) {
//             const namefam = cur_frm.doc.family_members[i]
//             console.log(namefam.name1);
// }
// },
//     age:function(frm) {
//         const age1 = frm.doc.age
//         if (age1 < 18) {
//           frappe.msgprint("vaa da chinna paiyale")
//         } else if (age1 => 18) {
//           frappe.msgprint("Vannako senior")
//         } else {
//           frappe.throw("Vannako senior")
//         }
// }






// frappe.ui.form.on('Client Side Scripting', {

	    
	    
// // message alert
// // 		frappe.msgprint("Vanakkam ");

// // error alert
// // frappe.throw("unna sodhikka poradhave da indha frappe veluu");

// // refresh
// refresh: function(frm) {
//     frappe.msgprint("refresh event");
//     frm.dashboard.add_comment("refresh event", "yellow");
// 	}
	
// // 	before_load
// // before_load: function(frm) {
// //     frappe.msgprint("before_load event test"");
// // 	}	

// // 	onload
// // onload: function(frm) {
// //     frappe.msgprint("onload event test");
// // 	}	

// // onload_post_render
// // onload_post_render: function(frm) {
// //     frappe.msgprint("onload_post_render event test");
// // 	}	

// // validate
// // validate: function(frm) {
// //     frappe.throw("validate event test");
// // 	}

// // before_save
// // before_save: function(frm) {
// //     frappe.throw("before_save event test");
// // 	}	

// // after_save
// // after_save: function(frm) {
// //     frappe.throw("after_save event test");
// // 	}

// // enable
// // enable: function(frm) {
// //     frappe.msgprint("enable event test");
// // 	}

// // before_submit
// // before_submit: function(frm) {
// //     frappe.throw("before_submit event test");
// // 	}	

// // on_submit
// // on_submit: function(frm) {
// //     frappe.msgprint("on_submit event test");
// // 	}

// // before_cancel
// // before_cancel: function(frm) {
// //     frappe.throw("before_cancel event test");
// // 	}	

// // after_cancel
// // after_cancel: function(frm) {
// //     frappe.msgprint("after_cancel event test");
// // 	}

// });

// // child table 
// frappe.ui.form.on('Family Member', {
    
// //   name1: function(frm){
// //      frappe.msgprint("child table event test");  
// //   } 

// age(frm,cdt,cdn){
//   frappe.msgprint("age event test");  
// }


// });



// Copyright (c) 2024, swetha and contributors
// For license information, please see license.txt

// frappe.ui.form.on('Server Side Scripting', {
//     // enable: function(frm){
// 	// 	frm.call({
// 	// 		doc:frm.doc,
// 	// 		method: 'frm.call',
// 	// 		arg:{
// 	// 			msg:'hello'
// 	// 		},
// 	// 		freeze:true,
// 	// 		freeze_message:_('calling method'),
// 	// 		callback: function(r){
// 	// 			frappe.msgprint(r.message)
// 	// 			// frappe.msgprint("server side calling")
// 	// 			// frm.refresh_field('medication_order')
// 	// 		}
// 	// 	})
// 	// }
	
// });
// frappe.ui.form.on('Server Side Scripting', {
// 	//   enable: function(frm){
// 	// 	frm.call({
// 	// 		doc:frm.doc,
// 	// 		method:'frm.call',
// 	// 		arg:{
// 	// 			msg:'hello'
// 	// 		},
// 	// 		freeze:true,
// 	// 		freeze_message:_('calling method'),
// 	// 		callback: function(r){
// 	// 			frappe.msgprint(r.message)
// 	// 			// frappe.msgprint("server side calling")
// 	// 			// frm.refresh_field('medication_order')
// 	// 		}
// 	// 	})
// 	// }
// });

File: basic_learning.py
# Copyright (c) 2024, swetha and contributors
# For license information, please see license.txt

  #   def validate(self):
  #      frappe.msgprint("Vannako da mappule")
     
  #  def before_save(self):
  #     frappe.throw("save pannilaamaa?")
    
  #before_insert
  #after_instert  - validate befor savaing data to db
  #on_update - trigger while updating anything
  #on_submit
  #on_trash - trigger at the time of deleting
  #after_delete - trigger after deleting

 # def validate(self):
 #      full_name = self.first_name+" "+self.middle_name+" "+self.last_name
 #      self.set("full_name",full_name)

 #      array_Of_Members = self.get("family_members")
 #      total_salary = 0
 #      for member in array_Of_Members:
 #       total_salary+=member.salary
 #      self.set('total_salary', total_salary)  

# get doc ############################ 
# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):

#   def validate(self):
      
#       array_Of_Members = self.get("family_members")
#       total_salary_server = 0
#       for member in array_Of_Members:
#        total_salary_server+=member.salary
#       self.set('total_salary_server', total_salary_server)  

#       self.get_document()
#       total = total_salary_server + client_side_salary_total
#       self.set("total_salary",total)
      
#   def get_document(self):
#       doc = frappe.get_doc("Client Side Scripting", self.client_side_scripting)
#       full_name = doc.full_name
#       frappe.msgprint(full_name)
#       self.set("full_name",full_name)
#       client_salary = doc.total_salary
#       # frappe.msgprint(str(client_salary))
#       self.set("client_side_salary_total",client_salary)

    # def validate(self):
    #  self.get_document()
    # def get_document(self):
    #      doc = frappe.get_doc("Client Side Scripting", self.client_side_scripting)
    #      for row in doc.get("family_members"):
    #          frappe.msgprint(row.name1)



# new doc ###########################################
# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):

#     def validate(self):
#      self.new_document()  

#     def new_document(self):
#        doc = frappe.new_doc("Client Side Scripting") 
#        doc.first_name= "alice"
#        doc.last_name="arpudha"
#        doc.age=13
      
#        doc.append("family_members",{ 
#           "name1":"sivaji",
#           "relation":"summa",
#           "age":12,
#           "salary":200
#          },
#          )
#        doc.insert()

 # delete doc ########################################### 
# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):

#     def validate(self):
#         frappe.delete_doc("Client Side Scripting","PR-0003")



#get_list ##################################################    
# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):    
    
  
#  def validate(self):
#   self.get_list()

#  def get_list(self):
#   doc = frappe.db.get_list('Client Side Scripting',
#                            filters={
#                             'enable': 1
#                            },
#                            fields=['age','first_name']
#                            ) 
#   for d in doc:
#    age=str(d['age'])
#    first_name= d['first_name'] if d['first_name'] else 'unknown'
#    frappe.msgprint(f"Age:{age}, first Name: {first_name}")

# get value #####################################################
# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):
    
#     def validate(self):
#         self.get_value()

#     def get_value(self):
#         doc = frappe.db.get_value('Client Side Scripting','PR-0011',['first_name','age'])

#         first_name, age = doc if doc else (None, None)
#         first_name= first_name if first_name else 'Unknown'
#         age= str(age) if age is not None else 'N/A'
#         frappe.msgprint(f"Age:{age}, first Name:{first_name}")

# set value ######################################################

# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):
    
#     def validate(self):
#         self.set_value()

#     def set_value(self):
#         frappe.db.set_value('Client Side Scripting','PR-0011','first_name', 'sam')
#         first_name, age = frappe.db.get_value('Client Side Scripting','PR-0011',['first_name','age'])
#         frappe.msgprint(f"first Name:{first_name}, Age:{age}")  

# get and set list of value ###########################################

# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):
    
#     def validate(self):
#         self.set_value()
#         self.get_value()

#     def set_value(self):
#       frappe.db.set_value('Client Side Scripting','PR-0011',{
#          "first_name":"sweamaa",
#          "age":25
#       })

#     def get_value(self):
#        doc = frappe.db.get_value('Client Side Scripting', 'PR-0011', ['first_name', 'age'])
#        first_name, age = doc if doc else (None, None) 
#        first_name = first_name if first_name else 'Unknown'
#        age = str(age) if age is not None else "N/A"
#        frappe.msgprint(f"Age:{age}, first Name:{first_name}")

# exists() ##################################################

# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):
    
#     def validate(self):
#         if frappe.db.exists('Client Side Scripting','PR-0011'):
#             frappe.msgprint("doc irukuu!!")
#         else :
#             frappe.msgprint("onnum illa")
        

# count() ########################################################
# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):
    
#     def validate(self):
#         frappe_doc= frappe.db.count('Client Side Scripting',{'enable':1})
#         frappe.msgprint(str(frappe_doc))


# sql() ###########################################################

# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):
    
#     def validate(self):
#         self.sql()

#     def sql(self):

#         data= frappe.db.sql("""
#                             SELECT
#                              first_name,
#                              age
#                             FROM
#                              `tabClient Side Scripting`
#                             WHERE
#                                enable = 1
#                             """, as_dict=1)    
#         for d in data:
#             # frappe.msgprint(f"First Name:{d['first_name']} Age:{str(d['age'])}")
#             ### or ###
#             frappe.msgprint(f"First Name:{d.first_name} Age:{str(d.age)}")


# frm.call() #####################################################################

# import frappe
# from frappe.model.document import Document

# class ServerSideScripting(Document):
#   pass
    # @frappe.whitelist()
    # def frm_call(self,msg):
    #     # import time
    #     # time.sleep(5)
    #     # frappe.msgprint(msg)

    #     # self.first_name='alvin'
    #     return "hIII"